### PR TITLE
Add ability to hide tabs on EntityLayout based on permissions

### PR DESCRIPTION
**Pull request by timbonicus**
Adds a `permission` prop on `EntityLayout.Route` to allow hiding EntityPage routes based on permissions. This could allow showing a tab only to the entity owners, for example. The rules around authorization of the permission are handled in the policy as usual.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/app/src/components/catalog/EntityPage.tsx [hunk 1]
```diff
@@ -485,7 +486,11 @@ const serviceEntityPage = (
       <EntityKafkaContent />
     </EntityLayout.Route>
 
-    <EntityLayout.Route path="/todos" title="TODOs">
+    <EntityLayout.Route
+      path="/todos"
+      title="TODOs"
+      permission={catalogEntityDeletePermission}
```
Comment (Rugvip) [inline packages/app/src/components/catalog/EntityPage.tsx:492]: Bit of a strange example this, is it not? 😅

File: plugins/permission-react/src/apis/IdentityPermissionApi.ts [hunk 2]
```diff
@@ -53,4 +53,11 @@ export class IdentityPermissionApi implements PermissionApi {
     );
     return response[0];
   }
+
+  async authorizeAll(requests: AuthorizePermissionRequest[]) {
```
Comment (Rugvip) [inline plugins/permission-react/src/apis/IdentityPermissionApi.ts:57]: What do you think about adding some lightweight same-tick caching to the `authorize` method instead of adding this? Feeling that that could bring bigger benefits on top of being able to batch requests from the same source, as that'd batch requests from separate sources too.

File: plugins/catalog/src/components/EntityLayout/EntityLayout.tsx [hunk 3]
```diff
@@ -185,33 +193,63 @@ export const EntityLayout = (props: EntityLayoutProps) => {
...
-          withStrictError:
-            'Child of EntityLayout must be an EntityLayout.Route',
-        })
-        .getElements<EntityLayoutRouteProps>() // all nodes, element data, maintain structure or not?
-        .flatMap(({ props: elementProps }) => {
-          if (!entity) {
-            return [];
-          } else if (elementProps.if && !elementProps.if(entity)) {
+
+  const routeElements = useElementFilter(children, elements =>
```
Comment (Rugvip) [inline plugins/catalog/src/components/EntityLayout/EntityLayout.tsx:197]: There's another option here that might be a bit trickier but waaaaaay more generic, which is to add support for the `permission` prop to `useElementFilter`. We could add it later too either way though

File: docs/permissions/custom-rules.md [hunk 4]
```diff
@@ -86,12 +86,12 @@ class TestPermissionPolicy implements PermissionPolicy {
       return createCatalogConditionalDecision(
         request.permission,
 -       catalogConditions.isEntityOwner(
--         user?.identity.ownershipEntityRefs ?? [],
+-         { claims: user?.identity.ownershipEntityRefs ?? [] },
```
Comment (Rugvip) [inline docs/permissions/custom-rules.md:89]: Potentially split these doc updates into a separate PR to ship them asap? 😁

**Review by Rugvip — CHANGES_REQUESTED** [review]
Ooh, interesting 😁

File: packages/app/src/components/catalog/EntityPage.tsx (hunk 1 above)
Comment (timbonicus) [reply packages/app/src/components/catalog/EntityPage.tsx:492]: Yes, definitely 😅 I was just using it for testing, I should remove it or come up with a better example.

File: plugins/catalog/src/components/EntityLayout/EntityLayout.tsx (hunk 3 above)
Comment (timbonicus) [reply plugins/catalog/src/components/EntityLayout/EntityLayout.tsx:197]: How do you think this would look? Something like this?

```ts
  const routeElements = useElementFilter(children, elements =>
    elements
      .selectByComponentData({
        key: dataKey,
        withStrictError: 'Child of EntityLayout must be an EntityLayout.Route',
      })
      .withPermissionResource({ entity })
      .getElements<EntityLayoutRouteProps>(),
  );
```

File: docs/permissions/custom-rules.md (hunk 4 above)
Comment (timbonicus) [reply docs/permissions/custom-rules.md]: Harry beat me to it in https://github.com/backstage/backstage/pull/14196 💨

File: plugins/catalog/src/components/EntityLayout/EntityLayout.tsx (hunk 3 above)
Comment (Rugvip) [reply plugins/catalog/src/components/EntityLayout/EntityLayout.tsx:197]: Perhaps, unless we just rely on the entity context instead

**Comment by simonstamm** [comment]
Hey @timbonicus, I also need to hide tabs based on specific permissions. The only way I currently found is doing it like:
```
const defaultEntityPage = (
  <RequirePermission
    permission={permission}
    errorPage={
      <EntityLayout>
        <EntityLayout.Route path="/allowed-path" title="Overview">
          {somecontent}
        </EntityLayout.Route>
      </EntityLayout>
    }
  >
    <EntityLayout>
      <EntityLayout.Route path="/allowed-path" title="Overview">
        {somecontent}
      </EntityLayout.Route>

      <EntityLayout.Route path="/path" title="Docs">
        {permissionedContent}
      </EntityLayout.Route>
    </EntityLayout>
  </RequirePermission>
);
```

I don't really like this way. Is there a better way to solve it since your PR hasn't been merged?

**Comment by painotpi** [comment]
Hey @Rugvip can this be re-opened? unless there is a workaround to have this functionality?

**Comment by Rugvip** [comment]
@backstage/permission-maintainers any thoughts on this PR? Something you want to continue?

**Comment by painotpi** [comment]
@Rugvip Any update here? or is there a workaround to get this working with the existing setup? What @simonstamm mentioned may work at an `EntityLayout` level but I was looking at granular control at the `EntityLayout.Route` level.
